packages/typescript-checker/src/typescript-checker.ts

```diff
@@ -33,176 +29,149 @@ export function create(injector: Injector<PluginContext>): TypescriptChecker {
-    if (parsedConfig.error) {
-      return content; // let the ts compiler deal with this error
-    } else {
-      for (const referencedProject of retrieveReferencedProjects(parsedConfig, path.dirname(fileName))) {
-        this.allTSConfigFiles.add(referencedProject);
-      }
-      return overrideOptions(parsedConfig, buildModeEnabled);
+    const mutantErrorRelationMap = await this.checkErrors(mutants, {}, this.tsCompiler.getFileRelationsAsNodes());
+    for (const [id, errors] of Object.entries(mutantErrorRelationMap)) {
+      result[id] = { status: CheckStatus.CompileError, reason: this.createErrorText(errors) };
     }
+
+    return result;
   }
 
   /**
-   * Resolves the task that is currently running. Will report back the check result.
+   * Creates groups of the mutants.
+   * These groups will get send to the check method.
+   * @param mutants All the mutants to group.
    */
-  private resolveCheckResult(): void {
-    if (this.currentErrors.length) {
-      const errorText = ts.formatDiagnostics(this.currentErrors, {
-        getCanonicalFileName: (fileName) => fileName,
-        getCurrentDirectory: process.cwd,
-        getNewLine: () => EOL,
-      });
-      this.currentTask.resolve({
-        status: CheckStatus.CompileError,
-        reason: errorText,
-      });
+  public async group(mutants: Mutant[]): Promise<string[][]> {
+    if (this.typeScriptCheckeroptions.typeScriptChecker.strategy === 'noGrouping') {
+      return mutants.map((m) => [m.id]);
     }
-    this.currentTask.resolve({ status: CheckStatus.Passed });
+    const nodes = this.tsCompiler.getFileRelationsAsNodes();
+
+    const mutantsOutSideProject = mutants.filter((m) => nodes.get(toPosixFileName(m.fileName)) == null).map((m) => m.id);
+    const mutantsToTest = mutants.filter((m) => nodes.get(toPosixFileName(m.fileName)) != null);
+
+    const groups = createGroups(mutantsToTest, nodes);
+    const sortedGroups = groups.sort((a, b) => b.length - a.length);
+    const result = mutantsOutSideProject.length ? [mutantsOutSideProject, ...sortedGroups] : sortedGroups;
+
+    this.logger.info(`Created ${result.length} groups with largest group of ${result[0].length} mutants`);
+    return result;
   }
 
-  /**
-   * Clear state between checks
-   */
-  private clearCheckState() {
-    while (this.currentErrors.pop()) {
-      // Idle
+  private async checkErrors(
+    mutants: Mutant[],
+    errorsMap: Record<string, ts.Diagnostic[]>,
+    nodes: Map<string, Node>
+  ): Promise<Record<string, ts.Diagnostic[]>> {
+    const errors = await this.tsCompiler.check(mutants);
+    const mutantsThatCouldNotBeTestedInGroups = new Set<Mutant>();
+
+    //If there is only a single mutant the error has to originate from the single mutant
+    if (errors.length && mutants.length === 1) {
+      errorsMap[mutants[0].id] = errors;
+      return errorsMap;
+    }
+
+    for (const error of errors) {
+      const nodeErrorWasThrownIn = nodes.get(error.file?.fileName ?? '');
```

Comment:
When is `file` or `file.fileName` `undefined`, and why is `''` a valid alternative?